### PR TITLE
Expose Transforms to Python Binding

### PR DESCRIPTION
**Pull request by sungwy**
Hi folks, here's my attempt at exposing the BucketTransform as a python binding for use by the PyIceberg project.

I've have built an Arrow Python interface with C++ before, but this is my first time working on Rust and this codebase. I'd love to get your feedback on how I should organize my code, and other best practices to apply here.

Implementation is based on: https://github.com/apache/arrow/commit/1d2b4a55770fa4dbe24959b3b40c745964c3184

EDIT: Expose all transforms

### CODE REVIEW

File: bindings/python/pyproject.toml [hunk 1]
```diff
@@ -31,8 +31,30 @@ classifiers = [
   "Programming Language :: Python :: 3.12",
 ]
 
-[project.optional-dependencies]
-test = ["pytest"]
+# Duplicate section required: https://github.com/python-poetry/poetry/pull/9135
```
Comment (sungwy) [inline bindings/python/pyproject.toml:34]: There's an open item to allow poetry to use [project] section instead of [tool.poetry]. I'm exploring using poetry for dependency management because even installing test dependencies is starting to feel a little strange with a simple pyproject.toml file, since we need to maintain a separate list of dependencies in `.github/workflows/bindings_python_ci.yml` as well as in bindings/python/pyproject.toml (there's no way to install just the optional dependencies using pip or maturin, it only supports editable install of the root project in addition to the extra dependencies)

We could also exploring using other python dependency managers as well.

Comment (sungwy) [reply bindings/python/pyproject.toml]: Taking my own advice - using pypa/hatch instead: https://hatch.pypa.io/latest/

File: .github/workflows/bindings_python_ci.yml [hunk 2]
```diff
@@ -78,6 +78,6 @@ jobs:
         shell: bash
         run: |
           set -e
-          pip install dist/pyiceberg_core-*.whl --force-reinstall
-          pip install pytest
-          pytest -v
+          pip install hatch==1.12.0
```
Comment (Xuanwo) [inline .github/workflows/bindings_python_ci.yml:81]: How about split workflow related changes to another PR?

File: .licenserc.yaml [hunk 3]
```diff
@@ -31,4 +31,6 @@ header:
     - '**/DEPENDENCIES.*.tsv'
     # Release distributions
     - 'dist/*'
+    # Generated content by poetry
+    - poetry.lock
```
Comment (Xuanwo) [inline .licenserc.yaml:35]: After switching to `hatch`, do we still need this?

File: bindings/python/src/lib.rs [hunk 4]
```diff
@@ -17,6 +17,9 @@
 
 use iceberg::io::FileIOBuilder;
 use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+
+mod transform;
```
Comment (Xuanwo) [inline bindings/python/src/lib.rs:22]: I like this idea

File: bindings/python/src/lib.rs [hunk 5]
```diff
@@ -26,6 +29,8 @@ fn hello_world() -> PyResult<String> {
 
 #[pymodule]
 fn pyiceberg_core_rust(m: &Bound<'_, PyModule>) -> PyResult<()> {
+    use transform::bucket_transform;
     m.add_function(wrap_pyfunction!(hello_world, m)?)?;
+    m.add_function(wrap_pyfunction!(bucket_transform, m)?)?;
```
Comment (Xuanwo) [inline bindings/python/src/lib.rs:34]: pyo3 supports modules, how about adding a `transform` module like this?

```rust
use pyo3::prelude::*;

#[pymodule]
fn pyiceberg_core_rust(m: &Bound<'_, PyModule>) -> PyResult<()> {
    let transform_module = PyModule::new_bound(parent_module.py(), "transform")?;
    child_module.add_function(wrap_pyfunction!(bucket_transform, &transform_module)?)?;
    parent_module.add_submodule(&transform_module)
}
```

File: bindings/python/src/transform.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    fn from(err: PyO3IcebergError) -> PyErr {
+        PyValueError::new_err(err.to_string())
+    }
+}
+
+fn to_rust_arrow_array(ob: PyObject, py: Python) -> PyResult<ArrayRef> {
+    // prepare a pointer to receive the Array struct
+    let (array, schema) = to_ffi(&ArrayData::new_empty(&arrow::datatypes::DataType::Null))
+        .map_err(PyO3ArrowError::from)?;
+    let array_pointer = &array as *const _ as uintptr_t;
```
Comment (Xuanwo) [inline bindings/python/src/transform.rs:107]: I'm not comfortable using libc directly. How about using `pyo3::ffi::Py_uintptr_t` instead?

File: bindings/python/src/transform.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+use arrow::{
+    array::{make_array, Array, ArrayData, ArrayRef},
+    error::ArrowError,
+    ffi::{from_ffi, to_ffi},
+};
+use libc::uintptr_t;
+use pyo3::{exceptions::PyOSError, exceptions::PyValueError, prelude::*};
+
+#[derive(Debug)]
+enum PyO3ArrowError {
```
Comment (Xuanwo) [inline bindings/python/src/transform.rs:32]: I believe most ffi related to arrow has been handled in [`arrow-rs/pyarrow`](https://github.com/apache/arrow-rs/blob/master/arrow/src/pyarrow.rs). Would you like to take a look? Maybe we can reuse them.

**Review by Xuanwo — COMMENTED** [review]
Thanks a lot for this work!

File: .github/workflows/bindings_python_ci.yml (hunk 2 above)
Comment (sungwy) [reply .github/workflows/bindings_python_ci.yml:81]: Agreed. I put up this PR as a follow up: https://github.com/apache/iceberg-rust/pull/572

**Comment by sungwy** [comment]
Thank you for the reviews @Xuanwo - I've taken all of your feedback and refactored the code to use the arrow::pyarrow crate

File: bindings/python/src/transform.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+use iceberg::Error;
+
+use arrow::{
+    array::{make_array, Array, ArrayData},
+};
+use arrow::pyarrow::{FromPyArrow, ToPyArrow};
+use pyo3::{exceptions::PyValueError, prelude::*};
+
+#[derive(Debug)]
+enum PyO3IcebergError {
```
Comment (Xuanwo) [inline bindings/python/src/transform.rs:30]: How about implementing a simple `fn to_py_err(err: iceberg::Error) -> PyError`. So we can use:

```rust
let bucket = create_transform_function(&Transform::Bucket(num_buckets)).map_err(to_py_err)?;
```

But no need for implementing `Display`, `From` by hand.

File: bindings/python/src/lib.rs [hunk 9]
```diff
@@ -25,7 +28,25 @@ fn hello_world() -> PyResult<String> {
...
+
+    module.add_wrapped(wrap_pyfunction!(bucket_transform))?;
+    Ok(())
+}
+
+#[pymodule]
+fn pyiceberg_core_rust(py: Python, m: &Bound<'_, PyModule>) -> PyResult<()> {
     m.add_function(wrap_pyfunction!(hello_world, m)?)?;
+
+    // https://github.com/PyO3/pyo3/issues/759
```
Comment (Xuanwo) [inline bindings/python/src/lib.rs:42]: I will explore better ways for this. Not a blocker of this PR.

**Review by Xuanwo — COMMENTED** [review]
Thanks! Most changes look good to me. Here are some nits.

File: bindings/python/src/transform.rs (hunk 8 above)
Comment (sungwy) [reply bindings/python/src/transform.rs:30]: That worked like a charm! Thanks @Xuanwo

**Comment by sungwy** [comment]
> Thanks! Most changes look good to me. Here are some nits.

Hi @Xuanwo - Given that there's an outstanding issue with the submodules, I decided to just implement a class and staticmethods instead so that we can make use of the usage pattern proposed in the tests. Please let me know what you think.

I've added the conversions for all of the existing transforms for completeness.

The test suite prioritizes verifying the integration, rather than the correctness of the transformed values which is already verified in the rust based test suites.

**Comment by sungwy** [comment]
> Thanks @sungwy for this pr, LGTM!

Thank you @liurenjie1024 !! 🙌

**Review by Xuanwo — APPROVED** [review]
Thank you, @sungwy a lot for the great work! Also thank @liurenjie1024's review

**Comment by sungwy** [comment]
Thank you @Xuanwo for all your help!
